File: Java/src/main/java/parrot/Parrot.java
package parrot;

public abstract class Parrot {

    public abstract double getSpeed();

    double getBaseSpeed() {
        return 12.0;
    }


    static Parrot european() {
        return new EuropeanParrot();
    }

    static Parrot african(int numberOfCoconuts) {
        return new AfricanParrot(numberOfCoconuts);
    }

    static Parrot norwegianBlue(double voltage, boolean isNailed) {
        if(isNailed){
            return new NailedParrot();
        }
        return new NorwegianBlueParrot(voltage);
    }

    static Parrot wasted() {
        return new WastedParrot();
    }
}

File: Java/src/main/java/parrot/Parrot.kt
package parrot

abstract class Parrot {

    abstract val speed: Double

    internal val baseSpeed: Double
        get() = 12.0

    companion object {

        fun european(): Parrot {
            return EuropeanParrot()
        }

        fun african(numberOfCoconuts: Int): Parrot {
            return AfricanParrot(numberOfCoconuts)
        }

        fun norwegianBlue(voltage: Double, isNailed: Boolean = false): Parrot {
            if (isNailed) {
                return NailedParrot()
            }
            return NorwegianBlueParrot(voltage)
        }

        fun wasted(): Parrot {
            return WastedParrot()
        }
    }
}